File: backend/src/main/java/com/spotscore/scoring/AttractivenessTier.java
package com.spotscore.scoring;

/**
 * 퍼센타일 밴드 등급(API 명세서 5.2.3절). percentileRank는 같은 industryCode 내에서
 * "낮을수록 상위"로 계산되므로(0=최상위), 등급 경계도 그 기준으로 나뉜다.
 *
 * 경계값(10/30/70)은 여기 한 곳에서만 관리한다 - 다른 곳에 매직 넘버로 중복하지 말 것.
 */
public enum AttractivenessTier {
    ATTRACTIVE, GOOD, AVERAGE, CAUTION;

    private static final double ATTRACTIVE_UPPER_BOUND_PERCENT = 10.0;
    private static final double GOOD_UPPER_BOUND_PERCENT = 30.0;
    private static final double AVERAGE_UPPER_BOUND_PERCENT = 70.0;

    public static AttractivenessTier fromPercentileRank(double percentileRank) {
        if (percentileRank <= ATTRACTIVE_UPPER_BOUND_PERCENT) {
            return ATTRACTIVE;
        }
        if (percentileRank <= GOOD_UPPER_BOUND_PERCENT) {
            return GOOD;
        }
        if (percentileRank <= AVERAGE_UPPER_BOUND_PERCENT) {
            return AVERAGE;
        }
        return CAUTION;
    }
}

File: frontend/src/styles/attractivenessTier.ts
import type { DefaultTheme } from 'styled-components';
import type { AttractivenessTier } from '../types/domain';

/**
 * percentileRank/attractivenessTier 표시용 매핑 (API 명세서 5.2.3절 퍼센타일 밴드).
 * 등급 경계값(10/30/70%) 자체는 백엔드 AttractivenessTier.java에서만 관리하고,
 * 여기는 이미 확정된 등급을 라벨/색으로만 바꾼다.
 */
export const ATTRACTIVENESS_TIER_LABEL: Record<AttractivenessTier, string> = {
  ATTRACTIVE: '매력적인 입지',
  GOOD: '괜찮은 입지',
  AVERAGE: '평균적인 입지',
  CAUTION: '신중한 검토 필요',
};

export const ATTRACTIVENESS_TIER_ICON: Record<AttractivenessTier, string> = {
  ATTRACTIVE: '🏆',
  GOOD: '✅',
  AVERAGE: '➖',
  CAUTION: '🔍',
};

/**
 * 여러 색을 섞은 신호등식 배색 대신, accent와 동일한 단일 색상 스케일
 * (theme.colors.scoreScale)에서 4단계만 골라 쓴다 — ATTRACTIVE가 가장 짙은
 * accent 그 자체, CAUTION은 위험 경고가 아니라 "재검토 권장" 톤이라 가장 옅은
 * 단계로 차분하게 표현한다.
 */
export function getAttractivenessTierColor(tier: AttractivenessTier, theme: DefaultTheme): string {
  const scale = theme.colors.scoreScale;
  switch (tier) {
    case 'ATTRACTIVE':
      return scale[4] ?? theme.colors.accent;
    case 'GOOD':
      return scale[3] ?? theme.colors.accent;
    case 'AVERAGE':
      return scale[2] ?? theme.colors.accent;
    case 'CAUTION':
      return theme.colors.textTertiary;
  }
}

export function formatPercentileLabel(percentileRank: number): string {
  const rounded = Math.round(percentileRank);
  return rounded <= 0 ? '상위 1% 이내' : `상위 ${rounded}%`;
}
